sdks/python/apache_beam/runners/common.py

```diff
@@ -159,6 +159,29 @@ def __init__(self, obj_to_invoke, method_name):
     self.args = args
     self.defaults = defaults
 
+    self.has_userstate_arguments = False
+    self.state_args_to_replace = {}
+    self.timer_args_to_replace = {}
+    for kw, v in zip(args[-len(defaults):], defaults):
+      if v.__class__ == core.DoFn.StateParam:
+        self.state_args_to_replace[kw] = v.state_spec
+        self.has_userstate_arguments = True
+      elif v.__class__ == core.DoFn.TimerParam:
+        self.timer_args_to_replace[kw] = v.timer_spec
+        self.has_userstate_arguments = True
+
+  def invoke_timer_callback(self, user_state_context, key, window):
+    # TODO(ccy): support WindowParam and TimestampParam.
```

Comment:
(can't reply to the original comment due to the rebase...)

Side inputs would be provided just as they are for process(...). This can remain a TODO.